### ISSUE TITLE
Passing authentication in the steps

### ISSUE THREAD

**Issue by germandb** [danielgerlag/workflow-core#111]
Hi Daniell, I'm having and issue passing my authentication class into the step definition. The problem that I have is the DI in the constructor of the step, my authentication class is registered as scope for my application and when the workflow core try to resolve the step, fail with this error

> Cannot resolve 'ADDInformatica.Cocktail.Workflows.Designer.Application.Activities.InitializeDataActivity' from root provider because it requires scoped service 'ADDInformatica.Cocktail.Core.Messaging.CorrelationTokenBus'.

This is the code of the step

`namespace ADDInformatica.Cocktail.Workflows.Designer.Application.Activities
{
    using System.Threading.Tasks;
    using ADDInformatica.Cocktail.Workflows.Designer.Application.Activities.MessagesDTOs;
    using Rebus.Bus;
    using WorkflowCore.Interface;
    using WorkflowCore.Models;

    /// <summary> Defines the activity that initializes the data of a workflow. </summary>
    public class InitializeDataActivity : StepBodyAsync
    {
        /// <summary>
        ///     Initializes a new instance of the <see cref="InitializeDataActivity"/> class.
        /// </summary>
        /// <param name="bus"> The bus. </param>
        public InitializeDataActivity(CorrelationTokenBus bus)
            : base(bus)
        {
        }

        /// <summary> Runs the activity that initialize the Workflow Entity. </summary>
        /// <param name="context"> The context. </param>
        /// <returns>
        ///     A task that contains a value indicating the execution result from the activity.
        /// </returns>
        public override async Task<ExecutionResult> RunAsync(IStepExecutionContext context)
        {
            InitializeDataActivityMessageDTO initializeDataActivityMessageDTO = new InitializeDataActivityMessageDTO();

            this.InitializeActivityMessageDTO(context, initializeDataActivityMessageDTO);

            initializeDataActivityMessageDTO.EntityName = this.EntityName;

            this.RunActivity(context, initializeDataActivityMessageDTO);

            return ExecutionResult.Next();
        }
    }
}`

The step is registry as Transient

**Comment by danielgerlag** [comment]
I think you also need to register any dependencies that are injected via the constructor

```

**Comment by germandb** [comment]
Hi Daniel I already have the TokenBus class registered as Scope `services.AddScope<CorrelationTokenBus>,`  I do that because the authentication is used for other parts of the application and dont work if I registered the class as Transient. I look into your code and saw that the WorkflowExecutor is registered as transient so I suppose that all the steps need to be registered transient to, this is correct?.
Can you think in a way to passe the authentication info if my class is registered as scope but the step is transient?.

**Comment by danielgerlag** [comment]
I don't think it matters that it is Transient or Scoped, but can you confirm they the workflow library and the rest of your application are sharing the same IoC container?  It sounds like you may have two containers?

**Comment by alexdrl** [comment]
The problem is with .NET Core 2.0 scope validation.
https://docs.microsoft.com/en-us/aspnet/core/fundamentals/hosting?tabs=aspnetcore2x#scope-validation

The problem is this case: When ValidateScopes is set to true, the default service provider performs checks to verify that:
Scoped services aren't directly or indirectly resolved from the root service provider.

It seems that Workflowcore uses the root service provider for resolving workflow steps and its dependencies, so no scoped service provider is created under workflow execution, or workflow resuming, this implies that from an activity we can not have a dependency to a Scoped service.

**Comment by danielgerlag** [comment]
You can't capture the scope of the original method that starts a workflow.  Each workflow is executed as a background process and could even be on a different machine, and in the case of long running workflows, could be months in the future.  You could attach some custom data to a workflow that is persisted along with it and use that to rebuild some context that your logic requires.

**Comment by germandb** [comment]
Hi Daniel, you were right, we are going to change the way our security it's handle and create a specific security token for the Workflows engine that get persisted with the steps and have the capability of impersonate the users when the steps get executed.
Thanks for your comments
